src/runtime/future/future.c

```diff
@@ -360,107 +240,140 @@ void future_register_callback(pony_ctx_t **ctx,
-
-  pony_gc_send(cctx);
-  trace_closure_entry(cctx, entry);
-  pony_send_done(cctx);
-
-  UNBLOCK;
+  closure_entry_t *entry = POOL_ALLOC(closure_entry_t);
+  *entry = (closure_entry_t){ .actor = (cctx)->current, .future = NULL, .closure = c };
+  chain_treiber_stack_push(&fut->chain_stack, entry);
+  if (__atomic_load_n(&(fut->fulfilled), __ATOMIC_SEQ_CST)) {
+    future_discharge(ctx, fut);
+  }
 }
 
+void future_await(pony_ctx_t **ctx, future_t *fut)
+{
+  if (__atomic_load_n(&(fut->fulfilled), __ATOMIC_SEQ_CST)) {
+    return;
+  }
+  ucontext_t uctx;
+  future_tnode_info_t * pony_node = POOL_ALLOC(future_tnode_info_t);
+  *pony_node = (future_tnode_info_t) { .fut = (void*)fut, .awaited_uctx = &uctx, .isget = false, .fclass = FUTURE };
+  actor_await(ctx, &uctx, (void *)pony_node);
+}
 
-static void future_block_actor(pony_ctx_t **ctx, future_t *fut)
+encore_arg_t future_get_actor(pony_ctx_t **ctx, future_t *fut)
 {
-  perr("future_block_actor");
-  pony_ctx_t* cctx = *ctx;
-  pony_actor_t *a = (cctx)->current;
-  BLOCK;
+  if (!__atomic_load_n(&(fut->fulfilled), __ATOMIC_SEQ_CST)) {
+    ENC_DTRACE2(FUTURE_BLOCK, (uintptr_t) *ctx, (uintptr_t) fut);
+    future_block_actor(ctx, fut);
+    ENC_DTRACE2(FUTURE_UNBLOCK, (uintptr_t) *ctx, (uintptr_t) fut);
+  }
+  ENC_DTRACE2(FUTURE_GET, (uintptr_t) *ctx, (uintptr_t) fut);
+  return fut->value;
+}
 
-  if (fut->fulfilled) {
-    UNBLOCK;
+static inline void future_block_actor(pony_ctx_t **ctx, future_t *fut)
+{
+  perr("future_block_actor");
+  if (__atomic_load_n(&(fut->fulfilled), __ATOMIC_SEQ_CST)) {
     return;
   }
-
+  pony_ctx_t* cctx = *ctx;
+  pony_actor_t *a = cctx->current;
   pony_unschedule(cctx, a);
-  assert(fut->no_responsibilities < 16);
-  fut->responsibilities[fut->no_responsibilities++] = (actor_entry_t) { .type = BLOCKED_MESSAGE, .message = (message_entry_t) { .actor = a } };
-
   encore_actor_t *actor = (encore_actor_t*) a;
-
-  assert(actor->lock == NULL);
-  actor->lock = &fut->lock;
-  actor_block(ctx, actor);
+  future_tnode_info_t * pony_node = POOL_ALLOC(future_tnode_info_t);
+  *pony_node = (future_tnode_info_t) { .fut = (void*)fut, .awaited_uctx = NULL, .isget = true, .fclass = FUTURE };
+  actor_block(ctx, actor, (void *)pony_node);          
 }
 
-// ===============================================================
-// Possibly these functions do not belong in the future library
-// ===============================================================
+void future_fulfil(pony_ctx_t **ctx, future_t *fut, encore_arg_t value) 
+{
+  assert(fut->fulfilled == false);
+  ENC_DTRACE2(FUTURE_FULFIL_START, (uintptr_t) *ctx, (uintptr_t) fut);
+  fut->value = value;
+  __atomic_store_n(&(fut->fulfilled), true, __ATOMIC_SEQ_CST);
+  future_gc_send_value(*ctx, fut);
+  future_discharge(ctx, fut);
+  ENC_DTRACE2(FUTURE_FULFIL_END, (uintptr_t) *ctx, (uintptr_t) fut);
+}
 
-void future_await(pony_ctx_t **ctx, future_t *fut)
+void future_discharge(pony_ctx_t **ctx, future_t *fut)
```

Comment:
I am not sure of what the naming means here. Maybe it's best if you provide some more information.
At the same time, I get the feeling that this *discharge* is possibly doing too many things. As I understood:
1. Goes through the list of blocked actors and schedules them
2. Starts running chained computations attached to the future, following a LIFO order. I would have expected that computations that are attached first run before the ones attached later (if the chaining operation happens in the same actor, and it's fair to assume any order if the future is shared among actors that may chain computations in this same future)